### PR TITLE
Improve revng-check-conventions script

### PR DESCRIPTION
**Pull request by mrjackv**

### CODE REVIEW

File: scripts/revng-check-conventions [hunk 1]
```diff
@@ -70,7 +73,9 @@ USE_CLANG_FORMAT_FILE="0"
 
 PRINT_CLANG_FORMAT_CONFIG="0"
 
-while [[ $# > 0 ]]; do
+GIT_LS_FILES="git ls-files -om --exclude-standard"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:76]: Drop `-o`, we only want to make changes to git-tracked files.

File: scripts/revng-check-conventions [hunk 2]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
     esac
 done
 
+# Retrive files by extension or shebang
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:112]:
```suggestion
# Retrieve files by extension or shebang
```

File: scripts/revng-check-conventions [hunk 3]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+        if head -n1 "$file" | grep -qE "^#!.*${2}"; then
+            shebang_files+=("$file")
+        fi
+    done
+    files=( "${ext_files[@]}" "${shebang_files[@]}")
+    if [[ ${#files[@]} -eq 0 ]]; then return; fi
+    printf "%s\n" "${files[@]}" | sort -u
+}
+
+python_files="find_by_ext_or_shebang py python"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:129]: This doesn't work with `python3`

File: scripts/revng-check-conventions [hunk 4]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
     esac
 done
 
+# Retrive files by extension or shebang
+# $1: extension to match
+# $2: shebang to match
+function find_by_ext_or_shebang() {
+    readarray -t ext_files < <($GIT_LS_FILES | grep "\.${1}$")
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:116]: * Please make all variable UPPERCASE.
* I'd honestly be happier to have a check on all `git ls-files` for the absence of space/newline and then just go the old way. I understand you want to go The Right Way™, but it's mostly OCD, if we have files with spaces or anything else, it's just a bug.

File: scripts/revng-check-conventions [hunk 5]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+    readarray -t dirty_files < <($GIT_LS_FILES)
+    shebang_files=()
+    for file in "${dirty_files[@]}"; do
+        if head -n1 "$file" | grep -qE "^#!.*${2}"; then
+            shebang_files+=("$file")
+        fi
+    done
+    files=( "${ext_files[@]}" "${shebang_files[@]}")
+    if [[ ${#files[@]} -eq 0 ]]; then return; fi
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:125]: Make multiline

File: scripts/revng-check-conventions [hunk 6]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+    printf "%s\n" "${files[@]}" | sort -u
+}
+
+python_files="find_by_ext_or_shebang py python"
+bash_files="find_by_ext_or_shebang sh bash"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:130]: Either call these `PYTHON_FILES_COMMAND` and `BASH_FILES_COMMAND`, or just invoke them right away (which I prefer).

File: scripts/revng-check-conventions [hunk 7]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
     esac
 done
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:111]: We're invoking `$GIT_LS_FILES` multiple times. We should do it once here.

File: scripts/revng-check-conventions [hunk 8]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+    FILES=("$@")
 fi
 
-GREP="git grep -n --color=always"
-
 # Run clang-format on FILES
 function run_clang_format() {
-    if test "$FORCE_FORMAT" -gt 0; then
-        clang-format -style="$CLANG_FORMAT_STYLE" -i $FILES || true
+    if [[ ${#FILES[@]} -eq 0 ]]; then return; fi
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:177]: Multiline

File: scripts/revng-check-conventions [hunk 9]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+    if [[ "$FORCE_FORMAT" -ne 0 ]]; then
+        cmake-format -i "${CMAKE_FILES[@]}" -l error
     else
-        cmake-format --check $(git ls-files | grep -i cmake) -l error
+        cmake-format --check "${CMAKE_FILES[@]}" -l error
+    fi
+}
+
+# Run shellcheck + check for set -euo pipefail
+function run_bash_check() {
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:208]: Nice! But I'm really reluctant to depend on Haskell...
How do you install this on Ubuntu? On Gentoo it's quite frustrating.

File: scripts/revng-check-conventions [hunk 10]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+        fi
+    done
+    if [[ ${#no_set_files[@]} -gt 0 ]]; then
+        echo "There are script files without 'set -euo pipefail'"
+        printf '%s\n' "${no_set_files[@]}"
     fi
 }
 
+# Check that all files have a license header
+function run_license_check() {
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:226]: Nice!

File: scripts/revng-check-conventions [hunk 11]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+
+
 # Run revng-specific checks on files
 function run_revng_checks() {
+    if [[ ${#FILES[@]} -eq 0 ]]; then return; fi
+    GREP="grep -Hn --color=always"
+
     # Check for lines longer than 80 columns
-    $GREP -E '^.{81,}$' $FILES | cat
+    $GREP -E '^.{81,}$' "${FILES[@]}" | cat_header "There are lines longer than 80 characters:"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:260]: Galaxy brain

File: scripts/revng-check-conventions [hunk 12]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
-        -e '^lib/Support/Assert\.cpp$' \
-        )"
+    readarray -t FILTERED_FILES < <(printf '%s\n' "${FILES[@]}" | \
+        grep -v -e '^runtime/support\.c$' -e '^lib/Support/Assert\.cpp$')
 
     for REGEXP in '\babort(' '\bassert(' 'assert(false' 'llvm_unreachable'; do
-        $GREP "$REGEXP" $FILTERED_FILES | cat
-    done
+        $GREP "$REGEXP" "${FILTERED_FILES[@]}"
+    done | cat_header "Assert code found where it should not be:"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:273]: More like "Use `revng_{assert,check,abort,unreachable}`"

File: scripts/revng-check-conventions [hunk 13]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
 
     # Things should never be at the end of a line
     for REGEXP in '::' '<' 'RegisterPass.*>' '} else' '\bopt\b.*>'; do
-        $GREP "$REGEXP\$" $FILES | cat
-    done
+        $GREP "$REGEXP\$" "${FILES[@]}"
+    done | cat_header "Things that shouldn't be at the end of a line:"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:278]: Ibidem. All so drop "things".

File: share/revng/pipelines/isolate-translate.yml [hunk 14]
```diff
@@ -1,3 +1,6 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
```
Comment (aleclearmind) [inline share/revng/pipelines/isolate-translate.yml:3]: Leave a blank line, everywhere.

File: scripts/revng-check-conventions [hunk 15]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+            head -n1 "$FILE" |  grep -E '^#pragma once$' > /dev/null || \
+                echo "$FILE: header does not start with #pragma once"
         fi
     done
 
 }
 
-if test $( ( run_clang_format && run_black && run_cmake_format && run_revng_checks | sort -u ) |& tee /dev/stderr | wc -c ) -ne 0; then
+tmp_file=$(mktemp)
+exec 1> >(tee "$tmp_file" > /dev/stdout)
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:333]: Comment what's going on here, very obscure.

File: scripts/revng-check-conventions [hunk 16]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
+# Cat, but print header if input is empty
+# stdin: lines to print
+# $1: Header to print if stdin is not empty
+function cat_header() {
+    local input
+    input="$(cat -)"
+
+    if echo "$input" | grep -vqz '^\s*$'; then
+        echo "$1"
+        echo "$input"
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:141]: Maybe intend with `sed /\n/\n  /`?

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: scripts/revng-check-conventions [hunk 17]
```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
...
     exit 0
 fi
 
 
 if [[ $# -eq 0 ]]; then
-    FILES="$(git ls-files | grep -E '\.(c|cc|cpp|h|hpp)$')"
+    readarray -t FILES < <($GIT_LS_FILES | grep -E '\.(c|cc|cpp|h|hpp)$')
 else
-    FILES="$@"
+    FILES=("$@")
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:172]: Please test that specifying individual files still works.

File: scripts/revng-check-conventions (hunk 1 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Done

File: scripts/revng-check-conventions (hunk 2 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: fixed

File: scripts/revng-check-conventions (hunk 15 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Added a comment to clarify

File: scripts/revng-check-conventions (hunk 16 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Fixed

File: share/revng/pipelines/isolate-translate.yml (hunk 14 above)
Comment (mrjackv) [reply share/revng/pipelines/isolate-translate.yml:3]: Fixed

File: scripts/revng-check-conventions (hunk 9 above)
Comment (mrjackv) [reply scripts/revng-check-conventions:225]: (Assuming > `18.04`)
`apt install shellcheck`
or (generally) unpack the binary archive from the [project's github release](https://github.com/koalaman/shellcheck/releases)

File: scripts/revng-check-conventions (hunk 12 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Done

File: scripts/revng-check-conventions (hunk 13 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Reworded all headers, they should sound better now.

File: scripts/revng-check-conventions (hunk 6 above)
Comment (mrjackv) [reply scripts/revng-check-conventions:130]: Moved all file list generation at the beginning

File: scripts/revng-check-conventions (hunk 7 above)
Comment (mrjackv) [reply scripts/revng-check-conventions:111]: Fixed

File: scripts/revng-check-conventions (hunk 5 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: Done

File: scripts/revng-check-conventions (hunk 4 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: For issue 1: All variables are uppercase now
For issue 2:
I've added a whitespace check before any file operations
I've tried converting everything to newline but a lot of things look uglier, with a lot `if [[ $(wc -l <<< "$FILES") -eq 0 ]]` and `tr '\n' ' '` that also spawn an extra process
also `${<var>[@]}` guarantees that each element is quoted before being passed as an argument, which is better
I've tried to get to a compromise by declaring all the `_FILES` variables together (to remove possible redundancies) and to have `ALL_FILES_N` and `FILES_N` (N is for newline) to allow easier `grep`-ing if needed

Moreover not quoting variables to explicitly allow splitting is [considered bad practice](https://github.com/koalaman/shellcheck/wiki/SC2086) by shellcheck, which would mean putting an ignore each time we invoke a function

File: scripts/revng-check-conventions (hunk 3 above)
Comment (mrjackv) [reply scripts/revng-check-conventions]: I've tried the following shebangs:
```
#!/usr/bin/env python3
#!/usr/bin/python3
```
and both are detected
Do you have a file to reproduce the issue?

File: scripts/revng-check-conventions [hunk 18]
```diff
@@ -100,134 +109,275 @@ while [[ $# > 0 ]]; do
...
+                echo "$FILE: header does not start with #pragma once"
         fi
     done
 
 }
 
-if test $( ( run_clang_format && run_black && run_cmake_format && run_revng_checks | sort -u ) |& tee /dev/stderr | wc -c ) -ne 0; then
+# Replace fd #1 (stdout) with process substitution
+# so that after the exec any write to it will also
+# get written into $TMP_FILE
```
Comment (aleclearmind) [inline scripts/revng-check-conventions:362]: Indent at 80

Comment (mrjackv) [reply scripts/revng-check-conventions]: Fixed

File: scripts/revng-check-conventions [hunk 19]
```diff
@@ -100,134 +109,275 @@ while [[ $# > 0 ]]; do
...
 # Run black
 function run_black() {
-    readarray -t PYTHON_FILES < <(( git grep -l '^#!.*python' && git ls-files | grep '\.py$' ) | sort -u )
     if [[ ${#PYTHON_FILES[@]} -eq 0 ]]; then return; fi
-    if test "$FORCE_FORMAT" -gt 0; then
+    if [[ "$FORCE_FORMAT" -ne 0 ]]; then
         black -q -l 100 "${PYTHON_FILES[@]}"
     else
-        black --check -l 100 "${PYTHON_FILES[@]}" | grep 'Oh no'
+        black --check -l 100 "${PYTHON_FILES[@]}"
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply scripts/revng-check-conventions]: done

[the remark this answers is not in the document]
Comment (mrjackv) [reply scripts/revng-check-conventions]: added

**Comment by aleclearmind** [comment]
Commits do not respect the 50/72 convention.

**Comment by mrjackv** [comment]
> Commits do not respect the 50/72 convention.

Fixed

File: python/revng/cli/revng.py [hunk 20]
```diff
@@ -10,10 +10,11 @@
 
 
 def extend_list(list, new_items):
-    new_items_set = set(new_items)
+    # new_items_set = set(new_items)
```
Comment (aleclearmind) [inline python/revng/cli/revng.py:13]: I see you're commenting code here and there.
We do not commit commented code. If it's dead code, drop it.

File: scripts/flake8-config [hunk 21]
```diff
@@ -0,0 +1,7 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+[flake8]
+max-line-length = 100
+extend-ignore = E203
```
Comment (aleclearmind) [inline scripts/flake8-config:7]: I see you're ignoring `F401` in several places, shall we put it here?
People looking at the code won't have any idea of what `F401` is or why it's silenced in this file and not in another.

File: tests/tuple-tree-generator/python-wrappers/multiple-versions/deserialize_multiple_versions.py [hunk 22]
```diff
@@ -40,9 +41,11 @@ def test_deserialize_multiple_versions():
...
 
 def test_tagged_deserialize_multiple_versions():
-    """Tests that the custom YAML loaders can be used to deserialize multiple conflicting versions at the same time.
-    Differing from test_deserialize_multiple_versions, this test uses tagged documents to test that the default path
-    resolver installed in the `YamlLoader`s does not cause issues.
+    """Tests that the custom YAML loaders can be used to deserialize multiple
+    conflicting versions at the same time.
+    Differing from test_deserialize_multiple_versions, this test uses tagged
+    documents to test that the default path resolver installed in the
+    `YamlLoader`s does not cause issues.
```
Comment (aleclearmind) [inline tests/tuple-tree-generator/python-wrappers/multiple-versions/deserialize_multiple_versions.py:48]: Seems like you're indenting at 80, but in `black` we force indentation to `100`.

File: scripts/monotone_framework_lattice.py [hunk 23]
```diff
@@ -224,7 +226,7 @@ def main():
                       - %isLessOrEqual% the C++ function with signature `bool isLessOrEqual(const LatticeElement &LHS, const LatticeElement &RHS)`
                       - %combineValues% the C++ function with signature `bool combineValues(const LatticeElement &LHS, const LatticeElement &RHS)`
                       - %transfer% the C++ function with signature `bool transfer(TransferFunction T, const LatticeElement &RHS)`
-                      """,
+                      """,  # noqa: E501
```
Comment (aleclearmind) [inline scripts/monotone_framework_lattice.py:229]: This kind of annotations are quite obscure, I think we should either disable the check in flake or add `noqa` and comment what's going on.
For this case (line too long), we might disable it. `black` should already to a good enough job in this sense.

File: scripts/monotone_framework_lattice.py [hunk 24]
```diff
@@ -24,17 +23,20 @@ def gen_combine_values(lattice, reachability):
...
     result = defaultdict(lambda: [])
     for v1, v1_data in lattice.nodes(data=True):
         for v2, v2_data in lattice.nodes(data=True):
             if v1 != v2:
                 i1 = int(v1_data["index"])
                 i2 = int(v2_data["index"])
-                nonzero = lambda i: set([x[0] for x in enumerate(reachability[i]) if x[1] != 0])
+
+                def nonzero(i):
+                    return set([x[0] for x in enumerate(reachability[i]) if x[1] != 0])
```
Comment (aleclearmind) [inline scripts/monotone_framework_lattice.py:39]: Why was that bad? It seems like pretty regular lambda. Are we banning lambda entirely?

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: scripts/monotone_framework_lattice.py (hunk 24 above)
Comment (mrjackv) [reply scripts/monotone_framework_lattice.py:39]: https://www.flake8rules.com/rules/E731.html
Lambda assignment is discouraged by pep8, inline is ok
Ok: `foo(lambda x: x+1)`
Not ok: `foo = lambda x: x+1`

File: scripts/flake8-config (hunk 21 above)
Comment (mrjackv) [reply scripts/flake8-config]: Switched to `pylint` with long-form error codes

File: tests/tuple-tree-generator/python-wrappers/multiple-versions/deserialize_multiple_versions.py (hunk 22 above)
Comment (mrjackv) [reply tests/tuple-tree-generator/python-wrappers/multiple-versions/deserialize_multiple_versions.py]: Fixed

File: python/revng/cli/revng.py (hunk 20 above)
Comment (mrjackv) [reply python/revng/cli/revng.py]: Dropped

**Comment by aleclearmind** [comment]
I'm getting this:
```
$ orc shell ./scripts/revng-check-conventions --all
include/revng/Pipeline/ContainerSet.h:100:41: warning: code should be clang-formatted [-Wclang-format-violations]
      Content[Name] = (*Factories[Name])(Name);
                                        ^
************* Module revng.cli.merge_dynamic.merge_dynamic
python/revng/cli/merge_dynamic/merge_dynamic.py:13:0: E0401: Unable to import 'elftools.elf.constants' (import-error)
python/revng/cli/merge_dynamic/merge_dynamic.py:14:0: E0401: Unable to import 'elftools.elf.enums' (import-error)
************* Module revng.cli.merge_dynamic.parsed_elf
python/revng/cli/merge_dynamic/parsed_elf.py:8:0: E0401: Unable to import 'elftools.elf.elffile' (import-error)
************* Module revng.cli.support
python/revng/cli/support.py:15:0: E0401: Unable to import 'elftools.elf.dynamic' (import-error)
python/revng/cli/support.py:16:0: E0401: Unable to import 'elftools.elf.elffile' (import-error)
python/revng/cli/support.py:110:19: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
************* Module revng.model.v1.__init__
python/revng/model/v1/__init__.py:5:15: E0001: invalid syntax (<unknown>, line 5) (syntax-error)
************* Module revng-model-compare
python/scripts/revng-model-compare:21:0: E0401: Unable to import 'grandiso' (import-error)
************* Module compile-time-constants
scripts/compile-time-constants.py:11:0: E0401: Unable to import 'elftools.elf.elffile' (import-error)
************* Module tuple_tree_generator.generators.python
scripts/tuple_tree_generator/tuple_tree_generator/generators/python.py:5:0: E0401: Unable to import 'black' (import-error)
python/revng/cli/support.py:110: error: X | Y syntax for unions requires Python 3.10
scripts/monotone_framework.py:32: error: Incompatible types in assignment (expression has type "None", variable has type Module)
```
My expectation is that all of the following commands are silent:
```
git rebase -x 'orc shell ./scripts/revng-check-conventions --all' develop
git rebase -x 'orc shell ./scripts/revng-check-conventions --all --force-format' develop
```

In order to do this, you probably need to squash `Make run_black return 0 if there are no errors` into the first commit.

**Comment by mrjackv** [comment]
Squashed, both `git rebase -x` return no output

**Comment by mrjackv** [comment]
Dropped `pylint`, now `flake8` is used plus some plugins
There is a [sister PR](https://github.com/revng/orchestra/pull/76) in orchestra for the extra commits

**Comment by aleclearmind** [comment]
Merged in 58c8f96. Thanks!

**Pull request closed without merging**